File: Tree/cartesian-tree.cpp
// verified with:
// https://judge.yosupo.jp/submission/109768
// https://atcoder.jp/contests/abc255/submissions/35950420

template<typename T>
struct cartesian_tree{
  vector<int> L,R,par;

  explicit cartesian_tree(int _n, vector<T> _a): n(_n),val(_a) {
    L.assign(n,-1);
    R.assign(n,-1);
    par.assign(n,-1);
    path.reserve(n);
    for (int i = 0; i < n; ++i){
      int ko = -1;
      while(!path.empty() && val[i] < val[path.end()[-1]]){
        ko = path.end()[-1];
        path.pop_back();
      }
      if (ko != -1){
        par[ko] = i;
      }
      if (!path.empty()) par[i] = path.end()[-1];
      path.pb(i);
    }
    for(int i = 0; i < n; ++i){
      if (par[i] == -1) {
        par[i] = i; // ここは根の親の定義に依存
        continue;
      }
      if (i < par[i]){
        L[par[i]] = i;
      }else{
        R[par[i]] = i;
      }
    }
  }
  
  private:
    const int n;
    const vector<T> val;
    vector<int> path;
};

File: Tree/cartesian-tree.py
# verified with:
# https://judge.yosupo.jp/submission/109770
# https://atcoder.jp/contests/abc255/submissions/35950398

from typing import List

class cartesian_tree:
	def __init__(self, n:int, val:List[int]) -> None:
		self.par = [-1]*n
		self.L = [-1]*n
		self.R = [-1]*n
		path = []
		for idx,i in enumerate(val):
			ko = -1
			while path and i < val[path[-1]]:
				ko = path.pop()
			if ko != -1:
				self.par[ko] = idx
			if path:
				self.par[idx] = path[-1]
			path.append(idx)
		for i in range(n):
			if self.par[i] == -1:
				self.par[i] = i # ここは根の親の定義に依存
				continue
			if i < self.par[i]:
				self.L[self.par[i]] = i
			else:
				self.R[self.par[i]] = i